library/src/permutation/device/instance_params.cpp

```diff
@@ -0,0 +1,73 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include "instance_params.hpp"
+
+namespace ck::tensor_operation::device::instance
+{
+    std::vector<hiptensor::Uid> getHashCodesWithAllInOutScalarPerVectorSeq(
+        hipDataType                  typeIn,
+        hipDataType                  typeOut,
+        hiptensorOperator_t          aOp,
+        hiptensorOperator_t          bOp,
+        hiptensor::PermutationOpId_t scale,
+        index_t                      numDim,
+        index_t                      blockSize,
+        index_t                      m0PerBlock,
+        index_t                      m1PerBlock,
+        index_t                      m0PerThread,
+        index_t                      m1PerThread,
+        std::pair<index_t, index_t>  threadClusterArrangeOrder)
+    {
+        std::vector<hiptensor::Uid> hashCodes;
+
+        // - IMPORTANT: keep the order of scalarPerVectorSeq from 16 to 0 so that the instance
+        //      with greater scalarPerVectorSeq will be chose.
+        // - scalarPerVectorSeq is 0 when it is CPU reference instance.
+        // - `hashCodes` may contain hash codes that not represent any instances. It is not a problem
+        //      since these hash codes will be ignored.
+        auto scalarPerVectorSeqs = std::vector<index_t>{16, 8, 4, 2, 1, 0};
+        for(auto scalarPerVectorSeq : scalarPerVectorSeqs)
+        {
+            hashCodes.push_back(hiptensor::Hash{}(typeIn,
+                                                  typeOut,
+                                                  aOp,
+                                                  bOp,
+                                                  scale,
+                                                  numDim,
+                                                  blockSize,
+                                                  m0PerBlock,
+                                                  m1PerBlock,
+                                                  m0PerThread,
+                                                  m1PerThread,
+                                                  threadClusterArrangeOrder.first,
+                                                  threadClusterArrangeOrder.second,
+                                                  scalarPerVectorSeq,
```

Comment:
Do both in and out have to be the same? Or is it interesting to have a combinatorial set for both in and out?

Reply:
Yeah, hiptensor supports combinations. 

![image](https://github.com/user-attachments/assets/ec3163b4-b1f3-4015-a9dd-5c702eb46523)

cuTensor supports even more. 

|typeA | typeB | typeScalar
-- | -- | --
CUDA_R_16F | CUDA_R_16F | CUDA_R_16F
CUDA_R_16F | CUDA_R_16F | CUDA_R_32F
CUDA_R_16F | CUDA_R_32F | CUDA_R_32F
CUDA_R_32F | CUDA_R_16F | CUDA_R_32F
CUDA_R_16BF | CUDA_R_16BF | CUDA_R_16BF
CUDA_R_16BF | CUDA_R_16BF | CUDA_R_32F
CUDA_R_32F | CUDA_R_32F | CUDA_R_32F
CUDA_R_64F | CUDA_R_64F | CUDA_R_64F
CUDA_R_32F | CUDA_R_64F | CUDA_R_64F
CUDA_R_64F | CUDA_R_32F | CUDA_R_64F
CUDA_C_32F | CUDA_C_32F | CUDA_C_32F
CUDA_C_64F | CUDA_C_64F | CUDA_C_64F
CUDA_C_32F | CUDA_C_64F | CUDA_C_64F
CUDA_C_64F | CUDA_C_32F | CUDA_C_64F